### ISSUE TITLE
Multiple recipes with same ingredient not incrementing shopping list quantities

### ISSUE THREAD

**Issue by jshank** [grocy/grocy#840]
Version 2.7.1

I have 2 recipes with Scallions, when adding both of them to the shopping list, only the first recipe adds a quantity of scallions.

**Comment by jshank** [comment]
It's not showing another entry either, the ingredient is ignored.

**Comment by jshank** [comment]
I just confirmed using [Grocy Demo](https://en.demo.grocy.info/shoppinglist) by adding the recipe for Sandwiches and Pizza, both which call for pack of cheese. Only one pack of cheese is added.

**Comment by berrnd** [comment]
Yes, because currently the missing amount is only checked on per recipe level, so when 1 pack is already on the shopping list, that's already fulfilled for the second recipe...

That's done here (the amount already on the shopping list is subtracted on line 41):
https://github.com/grocy/grocy/blob/98f214e9f14d5197f85aaefd38de856bff5da208/services/RecipesService.php#L30-L58

**Comment by jshank** [comment]
This is for 2 scallions on one recipe and 4 for another. I need a total of 6 but only 2 are added.

**Comment by Wordna1** [comment]
Is this related to the Recipe checkbox "Do not check against the shopping list when adding missing items to it"?  Would checking this box on all related recipes resolve this issue?  If so, is there a way to make this box checked by default for my installation?  I am curious why the default is "needed amount - stock amount - shopping list amount" rather than "needed amount - stock amount".

**Comment by berrnd** [comment]
Yep, `Do not check against the shopping list when adding missing items to it` checked will always put the needed amount onto the shopping list.

> "needed amount - stock amount - shopping list amount" rather than "needed amount - stock amount

The initial idea was to only have the needed amount on the shopping list, regardless of how often you click the "Add missing ingredients to the shopping list" button. Of course we can make this somehow configurable.

And each recipe is checked for it's own currently, so the needed amounts are never cumulative unless `Do not check against the shopping list when adding missing items to it` is checked.

**Comment by jshank** [comment]
Quick way to mass update `Do not check against the shopping list when adding missing items to it`

This is based on the linuxserver/grocy docker image:
```
$ docker exec -it grocy bash
apk add sqlite3
sqlite3 /config/data/grocy.db
update recipes set not_check_shoppinglist=1;
<ctrl-d>
exit
```

**Comment by Wordna1** [comment]
@jshank , I am hoping your mass update code will work as a work-around to resolve #925 .  Is there any way to make "Do not check against the shopping list when adding missing items to it" the default for new ingredients, maybe even enforce this as the only option?

**Comment by berrnd** [comment]
>  Is there any way to make "Do not check against the shopping list when adding missing items to it" the default for new ingredients

It's not an option per ingredient, but per recipe.
Possible via [custom JS](https://github.com/grocy/grocy#adding-your-own-css-or-js-without-to-have-to-modify-the-application-itself) - here is a snippet (for `/data/custom_js.html`) which does that:
```html
<script>
	if (Grocy.EditMode == "create")
	{
		$("#recipe-form #not_check_shoppinglist").prop("checked", true);
	}
</script>
```

**Comment by Wordna1** [comment]
> > Is there any way to make "Do not check against the shopping list when adding missing items to it" the default for new ingredients
> 
> It's not an option per ingredient, but per recipe.
> Possible via [custom JS](https://github.com/grocy/grocy#adding-your-own-css-or-js-without-to-have-to-modify-the-application-itself) - here is a snippet (for `/data/custom_js.html`) which does that:
> 
> ```
> <script>
> 	if (Grocy.EditMode == "create")
> 	{
> 		$("#recipe-form #not_check_shoppinglist").prop("checked", true);
> 	}
> </script>
> ```

Thank you @berrnd .  Can I also make the checkbox read-only?

**Comment by jshank** [comment]
Fantastic @berrnd, thank you! I seem to forget to check that every time I create a new recipe!

**Comment by berrnd** [comment]
> Can I also make the checkbox read-only?

Sure - extended snippet from above:
```
<script>
	if (Grocy.EditMode == "create")
	{
		$("#recipe-form #not_check_shoppinglist").prop("checked", true).attr("disabled", "");
	}
</script>
```

**Comment by Saturo78** [comment]
The checkbox has to be set in the recipe.
Are there further plans to put this checkpox to the general recipe settings?

**Comment by berrnd** [comment]
> Are there further plans to put this checkpox to the general recipe settings?

No. This here is about making it somehow possible that multiple recipes get their needed/missing ingredients checked cumulatively, nothing else.

[Here above](https://github.com/grocy/grocy/issues/840#issuecomment-825817624) is already a [custom JS](https://github.com/grocy/grocy#adding-your-own-css-or-js-without-to-have-to-modify-the-application-itself) snippet to always set that checkbox, if you want that.

**Comment by Saturo78** [comment]
> > Are there further plans to put this checkpox to the general recipe settings?
> 
> No. This here is about making it somehow possible that multiple recipes get their needed/missing ingredients checked cumulatively, nothing else.
> 
> [Here above](https://github.com/grocy/grocy/issues/840#issuecomment-825817624) is already a [custom JS](https://github.com/grocy/grocy#adding-your-own-css-or-js-without-to-have-to-modify-the-application-itself) snippet to always set that checkbox, if you want that.

Thanks, this helps for new recipes. For the recent 350 ones i have to set the checkbox by hand.

**Comment by berrnd** [comment]
> Thanks, this helps for new recipes. For the recent 350 ones i have to set the checkbox by hand.

Well, maybe a nice evening activity if you like something like that - by using the API or simply and directly updating the field in the database (also for that [here above is already an example](https://github.com/grocy/grocy/issues/840#issuecomment-662492742)) it's probably a 5 second action.

The SQL command would be `UPDATE recipes SET not_check_shoppinglist = 1 WHERE type = 'normal';` - open `data/grocy.db` in ["DB Browser for SQLite"](https://sqlitebrowser.org/), execute that there, done.

**Comment by Saturo78** [comment]
Perfect, everything done in just a few seconds.

**Comment by Saturo78** [comment]
Thinking longer about this brings me to the point that its not as easy as i thougt first.
The fact if there should be a check against the shoppinglist maybe depends more from the product than from the recipe.

Maybe the check should verify if the product on the shopping list available which means it is not reserved for any recipe.
If the check results in the product is available it should be reseved for that recipe.
This also would require that even when products are entered manually, an intended use is specified so that recipes can check whether the ingredients may be used or not.

**Comment by berrnd** [comment]
Exactly that's what this here is about - adds nothing new - nobody (at least I'm not) said that setting `not_check_shoppinglist` is the solution, it's something totally different.

**Comment by curtisbarnard** [comment]
I'm not having an issue with getting the items onto the shopping list however I would like to be able to see a sum total of each product rather than multiple lines. For instance I have multiple recipes that use garlic, can the below not be condensed to a single line displaying the total needed (3.28 ounces)?

![image](https://github.com/grocy/grocy/assets/100104319/fc822537-a893-4bd8-9850-83f89e2c71bf)

I've tried adding out of stock items to the shopping list via meal plan and the recipe page with the same results. I've also checked and unchecked the `Do not check against the shopping list when adding missing items to it` with no difference.

**Comment by berrnd** [comment]
@curtisbarnard That's off-topic here and already described in #1890 - please check existing requests and keep it to one topic per request, thanks.

**Comment by Scotison** [comment]
Am I missing something here? Using the "put missing products on shopping list" only adds a single instance of a product. eg if I have burgers 2 nights in a row each needing 500g of mince. Clicking that button regardless of the state of "Do not check against the shopping list when adding missing items to it" option on the recipe, only a single 500g will be added to the shopping list. Expected behavior would be to either add 2x 500g or 1000g to the shopping list. I have found that clicking add individually on each recipe on the meal planner correctly add the right amount of items to the shopping list. Only quality of life issue is that it doesn't add the same products quantity together.

**Comment by berrnd** [comment]
@Scotison please simply read through all what has been mentioned here and on the referenced other threads in the last 4 years. Stock fulfillment checking works on per recipe level. A recipe is a recipe, a meal plan item is also a recipe, a whole meal plan week (references the buttons in the "week header") is another recipe. That's why it works for "instances" and for all the others not. This Feature Request here is about all those others. If it's only a quality of life issue for you personally, that's perfect I guess.

**Comment by Scotison** [comment]
> @Scotison please simply read through all what has been mentioned here and on the referenced other threads in the last 4 years. Stock fulfillment checking works on per recipe level. A recipe is a recipe, a meal plan item is also a recipe, a whole meal plan week (references the buttons in the "week header") is another recipe. That's why it works for those "instances" and for all the others not. This Feature Request here is about all those others. If it's only a quality of life issue for you personally, that's perfect I guess.

On further review adding recipes one by one doesn't work either. If I am eating steak and chips Monday, Wednesday and Friday. When I click the add to shopping list on the individual instance of the steak and chips on Monday it adds only that require quantity and then greys out the add to stopping list for ALL other instances of the same recipe so I cannot then add the rest. This doesn't seem like intended behavior.

**Comment by berrnd** [comment]
> On further review adding recipes one by one doesn't work either.
> When I click the add to shopping list on the individual instance of the steak and chips on Monday it adds only that require quantity and then greys out the add to stopping list for ALL other instances of the same recipe

Once again, exactly what this Feature Request is about.

&nbsp;
> This doesn't seem like intended behavior.

Believe it or not, it is. Just once again, again: Stock fulfillment checking checks only a single recipe at a time, it cannot take multiple recipes into account. This Feature Request here asks to make exactly this somehow possible. If this is a key feature for your specific use case, Grocy doesn't fit as of today.

Please don't further spam this thread, more than everything on-topic has been said, talking about the same every other moth again and again doesn't make anything better or clearer, thanks a lot.

**Comment by berrnd** [comment]
> On further review adding recipes one by one doesn't work either.
> When I click the add to shopping list on the individual instance of the steak and chips on Monday it adds only that require quantity and then greys out the add to stopping list for ALL other instances of the same recipe

Just to extend on that, also once again, again: If you don't go through each recipe on [/recipes](https://demo-prerelease.grocy.info/recipes), but add your recipes to [/mealplan](https://demo-prerelease.grocy.info/mealplan) instead, then use the "Put missing onto the shopping list"-button on the top of that page (week header), Grocy sums up everything needed for that whole week just fine (since technically that whole meal plan week is internally just another recipe, as pointed out here multiple times) - that's how week meal planning is intended to work as of now, and it surprisingly works.

**Comment by Scotison** [comment]
> > On further review adding recipes one by one doesn't work either.
> > When I click the add to shopping list on the individual instance of the steak and chips on Monday it adds only that require quantity and then greys out the add to stopping list for ALL other instances of the same recipe
> 
> Just to extend on that, also once again, again: If you don't go through each recipe on [/recipes](https://demo-prerelease.grocy.info/recipes), but add your recipes to [/mealplan](https://demo-prerelease.grocy.info/mealplan) instead, then use the "Put missing onto the shopping list"-button on the top of that page (week header), Grocy sums up everything needed for that whole week just fine (since technically that whole meal plan week is internally just another recipe, as pointed out here multiple times) - that's how week meal planning is intended to work as of now, and it surprisingly works.

OMG. As you suspected. I am am a fool. Weeks shown on the meal planner start Sunday but the "Put missing onto the shopping list"-button on the top of that page (week header) adds Sunday to Saturday. eg it was only adding the first day that I was looking. Thanks for your assistance and feel free to purge these.

**Comment by berrnd** [comment]
> Weeks shown on the meal planner start Sunday but the "Put missing onto the shopping list"-button on the top of that page (week header) adds previous weeks to the Sunday visible

When searching the planet for "grocy first day of week", just slightly more than one thread will pop up and since I know reading through all that also won't work there, once again too:
- Triple check that your system (OS), PHP and client timezone / locale settings match (the practically normal case and a requirement to make any time related stuff for Grocy work)
- Triple check [those `config.php` settings](https://github.com/grocy/grocy/blob/5a2600209f38df1ae78f7689c9e0f45a7e82fea0/config-dist.php#L26-L38)

But finally enough noise here, please use the [r/grocy subreddit](https://www.reddit.com/r/grocy) (as [mentioned in README](https://github.com/grocy/grocy?tab=readme-ov-file#questions--help--bug-reports--feature-requests)) for any further general usage questions.

**Comment by NebuPookins** [comment]
The fact that this bug keeps getting reported over and over again is probably indicative that the current UI design is unclear or unintuitive.

I think it's worth brainstorming what could be done to improve the UI in this respect. I leave it up to your decision as to whether you want that brainstorming to occur as part of this issue, or whether you want to have a separate issue for the UI clarity vs the underlying functionality.

Some ideas off the top of my head for UI clarity:

Instead of having a true/false checkbox for adding ingredients, have 3 values (perhaps represented as 3 columns): "Don't Add" (which is equivalent to today's behavior of not checking the checkbox), "Add if not already in cart" (which is equivalent to today's behavior of checking the checkbox) and "Add" (which is the behavior that everyone who reports this bug seems to want).

![Image](https://github.com/user-attachments/assets/2bd080da-e4d1-42d8-a665-29e1d7e9b3f7)

**Comment by berrnd** [comment]
> Do you want the brainstorming to occur as part of this issue

Since everything about this here is still crystal clear, nothing needs to be brainstormed at all (whatever that should bring, why not just do it) and some hundred more comments create the clutter we all need: Not here.

&nbsp;
>  or would you rather have a separate issue for the UI clarity

Nothing is currently broken on that topic and even if you probably found Grocy yesterday, not every little detail, which is not supa dupa crystal clear to you on a first sight, is the next big bug.

**Comment by NebuPookins** [comment]
Sorry, I'm not sure I understand.

It seems like this issue is reported repeatedly by many people. In #2753 , you mentioned that this same issue has been reported in #402, #1622, #1621, #925, #1848, #1789, #1965, #2020, #2174, #2418, #2626 and #2695.

Isn't this indicative of the fact that a lot of people find the current UI design to be confusing?

**Comment by ch-world** [comment]
As this is currently an important topic for me I decided to implement it. Normally I would raise a PR but as they are disabled not sure if this is welcomed.
Here is my suggestion how this could look like: https://github.com/grocy/grocy/compare/master...ch-world:grocy:feature/cumulative_amounts

* Add not_check_stock_fulfillment to the recipes_pos_resolved view
* Iterate over the ingredients and build the sum before substracting the amount in stock

**Comment by berrnd** [comment]
> Normally I would raise a PR but as they are disabled not sure if this is welcomed.

You just confirmed one of my reasons why I disabled them: Not doing end to end tests (or obviously not a single one at all), leading to more than a lot of wasted time on my end.

Any "Put missing products on shopping list"-button gets disabled when the amount needed is already on the shopping list. No eggs in stock, recipe x1 needs 1, put that on the shopping list, recipe x2 needs also one => button is disabled (due to already 1 on the shopping list) and I wonder how you even confirmed what you suggested is working for the most basic example. And no, please not go on with that here (ref [Contributing](https://github.com/grocy/grocy?tab=contributing-ov-file#readme)).
